impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -336,7 +339,10 @@ public void remove(final K key) throws CacheAccessException {
           }
 
           returnValue.set(mappedValue);
-          setAccessTimeAndExpiry(key, mappedValue, now);
+          if (setAccessTimeAndExpiry(key, mappedValue, now) == null) {
```

Comment:
Quite a number of the changes here are the same. Refactor into a new method like `setAccessTimeAndExpiryThenReturnMapping` which clarifies intent and includes expiration notification when needed.
Then you can have the few places where it happens differently still invoke a refactored `setAccessTimeAndExpiry`.